### ISSUE TITLE
IDEA: Embed cost sensors into the devices they're calculated from

### ISSUE THREAD

**Issue by whi-tw** [martinarva/dynamic_energy_cost#14]
Not doing the template for this one, sorry! I couldn't work out how to structure it into the headings!

I use https://github.com/bramstroker/homeassistant-powercalc for some devices (virtual power meter for devices, calculated by setting on, off, idle power draw readings).

This addon 'embeds' the energy / power sensors 'into' the original device:

<img width="503" alt="image" src="https://github.com/martinarva/dynamic_energy_cost/assets/3595020/9005470c-87e6-46e3-913b-1428c1e4fb84">

It's a super clean way of presenting this kind of information, as it reduces the number of individual devices on the system, and puts the energy information directly within the device, so it's available immediately when looking at the device.

Unsure of your thoughts on this, it could be complicated - I'm not 100% on the actual implementation details from powercalc.

I just wanted to draw attention to it as you may not have seen this pattern before, and it could be a good fit.

Unsure.

**Comment by martinarva** [comment]
Yep, this is in the roadmap. I have already checked the powercalc code, but was a bit overwhelming.

**Comment by martinarva** [comment]
I just can't get it to work. Must be something like this:


```Python
    @property
    def device_info(self):
        """Return device information to link this sensor with the integration or a specific device."""
        entity_registry = er.async_get(self.hass)
        entity_entry = entity_registry.async_get(self._energy_sensor_id)
        
        if entity_entry and entity_entry.device_id:
            # If there is a device linked to the user-inputted energy sensor, link the sensor to it.
            return {
                "identifiers": {(DOMAIN, self._energy_sensor_id)},  # Unique identifier for your sensor
                "name": self._device_name,
                "via_device": entity_entry.device_id  # Link to the device of the original energy sensor
            }
        else:
            # Fallback if no device is linked to the energy sensor
            return {
                "identifiers": {(DOMAIN, self._energy_sensor_id)},
                "name": self._device_name,
                "manufacturer": "Custom Integration"
            }
   ```


But it's not working

**Comment by TimSoethout** [comment]
Cool idea.
The Device Tools (https://community.home-assistant.io/t/device-tools-create-modify-and-merge-devices/691172) plugin allows you to do this manually by combining devices/entities etc.
However, it would be great if this was natively supported by Dynamic Energy Costs.

(Also maybe that code base can be used as inspiration?)
